008-migration.md

```diff
@@ -0,0 +1,83 @@
+# Migration
+
+- Implementation Owner: @torstendittmann
+- Start Date: (today's date, 13-01-2021)
+- Target Date: (expected date of completion, 13-01-2021)
+- Appwrite Issue:
+  [Is this RFC inspired by an issue in appwrite](https://github.com/appwrite/appwrite/issues/)
+
+## Summary
+
+[summary]: #summary
+
+Add a migration script to handle internal changes when upgrading to a new version with breaking changes.
+
+## Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+**What problem are you trying to solve?**
+
+During development, internal breaking changes are often introduced, which must be incorporated when upgrading to a new version. Also, the current migration script is not ready for easy extension of upcoming versions.
+
+**What is the context or background in which this problem exists?**
+
+Since Appwrite is still in early development, breaking change will most likely be introduced with each release. We will face a similar problem with major releases in the future.
+
+**Once the proposal is implemented, how will the system change?**
+
+Not much for the user, only for the developer. The concept of object-oriented programming is introduced to the migration script. Object Inheritance and Polymorphism via Interfaces will ensure easy and safe introduction of upcoming migrations of versions.
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+Initial situation is as follows, the current migration script serves its purpose - but is very inflexible and offers little basis for upcoming additions. To improve this, a main parent class and an interface is to be created, which is extended and implemented by the individual migrations. Significant refactoring is performed for the previous migrations.
+
+### New class and interface
+
+A new class will be introduced to `src/Appwrite/Migration`. The class will inherit instances (`$register`, `$projectDB` & co) and methods to the individual migrations, which are needed regardless of the version. Also, abstract functions are introduced which enforces certain methods that have to be custom to the migration. The main class will not be instantiable.
+
+It will look something like this:
+
+```php
+abstract class Migration {
+  private function __construct() { }
+
+  public function setProject($project, $projectDB) { /* ... */ }
+  public function forEachDocument($callback) { /* ... */ }
+  
+  abstract protected function execute();
+}
+
+class V07 extends Migration {
+    public function execute() { /* ... */ }
+}
+```
+
+Following functions are gonna be added the the main `Appwrite\Migration\Migration` class:
+
+#### `setProject(Document $project, Database $projectDB)`
+Sets the current project for migration. Will be used by the migration script run from the CLI.
```

Comment:
One thing I don't understand, once a user upgrades to newer version, aren't they suppose to migrate their all the projects, will it work on a project by project basis?

Reply:
Yes, the `setProject` is just for the class to set namespace & co to know which database. The migration script will loop throug all projects 👍🏻